CMake: Update GCC on macOS Conflict message

ax3l opened the pull request:
The latest release of CMake updates libuv, which fixes ObjC code usage on macOS. Passing ObjC code to non apple-clang compilers crashed the build before #16249.

But there are further issues to address before this conflict can be removed.

Refs.:
- https://gitlab.kitware.com/cmake/cmake/-/issues/20620 (fixed)
- https://gitlab.kitware.com/cmake/cmake/-/merge_requests/4687 (merged)
- https://gitlab.kitware.com/cmake/cmake/-/issues/21135 (open)

ax3l committed:
CMake 3.18.0+: Builds with GCC on macOS

The latest release of CMake updates libuv, which fixes ObjC code
usage on macOS. Passing ObjC code to non apple-clang compilers
crashed the build before.

Refs.:
- https://gitlab.kitware.com/cmake/cmake/-/issues/20620
- https://gitlab.kitware.com/cmake/cmake/-/merge_requests/4687

alalazo commented:
Just launched an installation of this spec:
```console
(venv) culpo@MacBook-Pro spack % spack spec cmake %gcc
Input spec
--------------------------------
cmake%gcc

Concretized
--------------------------------
cmake@3.18.2%gcc@9.2.0~doc+ncurses+openssl+ownlibs~qt arch=darwin-catalina-cannonlake
    ^ncurses@6.2%gcc@9.2.0~symlinks+termlib arch=darwin-catalina-cannonlake
        ^pkgconf@1.7.3%gcc@9.2.0 arch=darwin-catalina-cannonlake
    ^openssl@1.1.1g%gcc@9.2.0+systemcerts arch=darwin-catalina-cannonlake
        ^perl@5.30.3%gcc@9.2.0+cpanm+shared+threads arch=darwin-catalina-cannonlake
            ^berkeley-db@18.1.40%gcc@9.2.0 arch=darwin-catalina-cannonlake
            ^gdbm@1.18.1%gcc@9.2.0 arch=darwin-catalina-cannonlake
                ^readline@8.0%gcc@9.2.0 arch=darwin-catalina-cannonlake
        ^zlib@1.2.11%gcc@9.2.0+optimize+pic+shared arch=darwin-catalina-cannonlake
```

alalazo commented:
Using a Spack built GCC the build still errors out:
```
10 errors found in build log:
     1896    [ 58%] Building CXX object Source/CMakeFiles/CMakeLib.dir/LexerParser/cmDependsJavaParser.cxx.o
     1897    [ 58%] Building CXX object Source/CMakeFiles/CMakeLib.dir/LexerParser/cmCommandArgumentParser.cxx.o
     1898    [ 58%] Building CXX object Source/CMakeFiles/CMakeLib.dir/LexerParser/cmFortranLexer.cxx.o
     1899    [ 58%] Building CXX object Source/CMakeFiles/CMakeLib.dir/LexerParser/cmExprParser.cxx.o
     1900    In file included from /Users/culpo/PycharmProjects/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.0/gcc-9.2.0-ie4mp5d7u3ygeu3icgelktq437acyzw6/include/c
             ++/9.2.0/stdlib.h:36,
     1901                     from /private/var/folders/gy/g0z3m6dn0nz53nsttz2q0flm0000gn/T/culpo/spack-stage/spack-stage-cmake-3.18.2-5ylrs2biuqeuiwpjd4fm2osnhdqd45vm/spack-
             src/Source/LexerParser/cmFortranLexer.cxx:232:
  >> 1902    /Users/culpo/PycharmProjects/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.0/gcc-9.2.0-ie4mp5d7u3ygeu3icgelktq437acyzw6/include/c++/9.2.0/cstdlib:132:1
             1: error: '::aligned_alloc' has not been declared
     1903      132 |   using ::aligned_alloc;
     1904          |           ^~~~~~~~~~~~~
  >> 1905    make[2]: *** [Source/CMakeFiles/CMakeLib.dir/LexerParser/cmFortranLexer.cxx.o] Error 1
     1906    make[2]: *** Waiting for unfinished jobs....
     1907    In file included from /Users/culpo/PycharmProjects/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.0/gcc-9.2.0-ie4mp5d7u3ygeu3icgelktq437acyzw6/include/c
             ++/9.2.0/stdlib.h:36,
     1908                     from /private/var/folders/gy/g0z3m6dn0nz53nsttz2q0flm0000gn/T/culpo/spack-stage/spack-stage-cmake-3.18.2-5ylrs2biuqeuiwpjd4fm2osnhdqd45vm/spack-
             src/Source/LexerParser/cmExprLexer.cxx:232:
  >> 1909    /Users/culpo/PycharmProjects/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.0/gcc-9.2.0-ie4mp5d7u3ygeu3icgelktq437acyzw6/include/c++/9.2.0/cstdlib:132:1
             1: error: '::aligned_alloc' has not been declared
     1910      132 |   using ::aligned_alloc;
     1911          |           ^~~~~~~~~~~~~
     1912    In file included from /Users/culpo/PycharmProjects/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.0/gcc-9.2.0-ie4mp5d7u3ygeu3icgelktq437acyzw6/include/c
             ++/9.2.0/stdlib.h:36,
     1913                     from /private/var/folders/gy/g0z3m6dn0nz53nsttz2q0flm0000gn/T/culpo/spack-stage/spack-stage-cmake-3.18.2-5ylrs2biuqeuiwpjd4fm2osnhdqd45vm/spack-
             src/Source/LexerParser/cmCommandArgumentLexer.cxx:232:
  >> 1914    /Users/culpo/PycharmProjects/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.0/gcc-9.2.0-ie4mp5d7u3ygeu3icgelktq437acyzw6/include/c++/9.2.0/cstdlib:132:1
             1: error: '::aligned_alloc' has not been declared
     1915      132 |   using ::aligned_alloc;
     1916          |           ^~~~~~~~~~~~~
     1917    In file included from /Users/culpo/PycharmProjects/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.0/gcc-9.2.0-ie4mp5d7u3ygeu3icgelktq437acyzw6/include/c
             ++/9.2.0/stdlib.h:36,
     1918                     from /private/var/folders/gy/g0z3m6dn0nz53nsttz2q0flm0000gn/T/culpo/spack-stage/spack-stage-cmake-3.18.2-5ylrs2biuqeuiwpjd4fm2osnhdqd45vm/spack-
             src/Source/LexerParser/cmDependsJavaLexer.cxx:232:
  >> 1919    /Users/culpo/PycharmProjects/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.0/gcc-9.2.0-ie4mp5d7u3ygeu3icgelktq437acyzw6/include/c++/9.2.0/cstdlib:132:1
             1: error: '::aligned_alloc' has not been declared
     1920      132 |   using ::aligned_alloc;
     1921          |           ^~~~~~~~~~~~~
  >> 1922    make[2]: *** [Source/CMakeFiles/CMakeLib.dir/LexerParser/cmExprLexer.cxx.o] Error 1
  >> 1923    make[2]: *** [Source/CMakeFiles/CMakeLib.dir/LexerParser/cmCommandArgumentLexer.cxx.o] Error 1
  >> 1924    make[2]: *** [Source/CMakeFiles/CMakeLib.dir/LexerParser/cmDependsJavaLexer.cxx.o] Error 1
  >> 1925    make[1]: *** [Source/CMakeFiles/CMakeLib.dir/all] Error 2
  >> 1926    make: *** [all] Error 2

See build log for details:
  /var/folders/gy/g0z3m6dn0nz53nsttz2q0flm0000gn/T/culpo/spack-stage/spack-stage-cmake-3.18.2-5ylrs2biuqeuiwpjd4fm2osnhdqd45vm/spack-build-out.txt
```
even though the error seem a lot more like [this](https://stackoverflow.com/a/52777686/771663) than previous Objective C issue.

gartung commented:
I don't get the conflict error when I try `spack build cmake % gcc` on Catalina with Homebrew installed gcc@9.3.0

gartung commented:
But I do get the conflict error with cmake@3.17.3

```
 spack install cmake@3.17.3%gcc
==> Error: Conflicts in concretized spec "cmake@3.17.3%gcc@9.3.0~doc+ncurses+openssl+ownlibs~qt patches=e51119dd387aa3eb2f21fee730d3aa8b5480301c5e2e3a33bd86f4f7d8903861 arch=darwin-catalina-skylake/ahmkxje"
List of matching conflicts for spec:

    cmake@3.17.3%gcc@9.3.0~doc+ncurses+openssl+ownlibs~qt patches=e51119dd387aa3eb2f21fee730d3aa8b5480301c5e2e3a33bd86f4f7d8903861 arch=darwin-catalina-skylake
        ^ncurses@6.2%gcc@9.3.0~symlinks+termlib arch=darwin-catalina-skylake
            ^pkgconf@1.7.3%gcc@9.3.0 arch=darwin-catalina-skylake
        ^openssl@1.1.1g%gcc@9.3.0+systemcerts arch=darwin-catalina-skylake
            ^perl@5.30.3%gcc@9.3.0+cpanm+shared+threads arch=darwin-catalina-skylake
                ^berkeley-db@18.1.40%gcc@9.3.0 arch=darwin-catalina-skylake
                ^gdbm@1.18.1%gcc@9.3.0 arch=darwin-catalina-skylake
                    ^readline@8.0%gcc@9.3.0 arch=darwin-catalina-skylake
            ^zlib@1.2.11%gcc@9.3.0+optimize+pic+shared arch=darwin-catalina-skylake

1. "@:3.17.99 %gcc +ownlibs platform=darwin" conflicts with "cmake" [CMake<3.18.0 does not compile with GCC on macOS, use %apple-clang.]
```

alalazo commented:
@gartung If you can confirm a correct build of `cmake@3.18.0:` with Homebrew GCC I'll be merging this

gartung commented:
I get the same error as @alalazo with cmake@3.18.2 and Homebrew installed gcc@9.3.0

ax3l commented:
Ok, I will report another upstream issue and update the comment of the conflict (but keep it in).

Update: https://gitlab.kitware.com/cmake/cmake/-/issues/21135

ax3l commented:
Thank you both!
Can someone cross-check please if that also occurs with a GCC from Homebrew or if this might be related to a GCC bootstrapped with Spack from AppleClang only? It's curious, since a simple [`#include <stdlib.h>`](https://gitlab.kitware.com/cmake/cmake/-/blob/v3.18.2/Source/LexerParser/cmFortranLexer.cxx#L232) causes this.

gartung commented:
I used HomeBrew installed gcc@9.3.0

ax3l committed:
CMake: Further issues GCC+macOS

There are further issues to fix before this will work.
https://gitlab.kitware.com/cmake/cmake/-/issues/21135

ax3l commented:
Thanks a lot, updated the conflict - this is now ready to merge.

ax3l commented:
@gartung @alalazo are you able to compile this with your macOS + GCC?
```C++
// main.cpp
#include <cstdlib>

int main (){}
```
```bash
g++ -std=c++17 main.cpp
```
?

I think you might need to report a GCC upstream follow-up bug to https://gcc.gnu.org/bugzilla/show_bug.cgi?id=83662 for `cstdlib` on macOS (likely a GCC 7.4+ issue with C++17 and newer).

ax3l commented:
Sorry, I mean `#include <stdlib.h>`... Oh, that could be the bug in CMake...

alalazo commented:
I was about to say "I'm able to compile it"

ax3l commented:
and does
```C++
#include <stdlib.h>

int main (){}
```
crash for you with C++17?

alalazo commented:
Just tried and it seems to work. Used this snippet:
```c++
#include <stdlib.h>
#include <cstdio>

int main(){
  printf("Hello World\n");
}
```
compiled and run like:
```console
% /Users/culpo/PycharmProjects/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.0/gcc-9.2.0-ie4mp5d7u3ygeu3icgelktq437acyzw6/bin/g++ -std=c++17 main.cpp

% ./a.out                                                                                                                                                      
Hello World
```

ax3l commented:
Interesting... CMake includes a few more in [the lexer file that fails](https://gitlab.kitware.com/cmake/cmake/-/blob/v3.18.2/Source/LexerParser/cmFortranLexer.cxx#L232):
```C++
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <stdlib.h>

int main(){}
```
does that build with C++17, too? Something, something flex...

alalazo commented:
Yes, it does build. I see though that C standard library header files are not included as first thing in that file and that a lot of macro magic is probably happening in the preamble.

alalazo commented:
@ax3l The minimal snippet that reproduces the error:
```c++
#if !defined(_WIN32) && !defined(__sun)
/* POSIX APIs are needed */
#  define _POSIX_C_SOURCE 200809L
#endif

#include <stdlib.h>

int main(){}
```

gartung commented:
```
/usr/local/Cellar/gcc/9.3.0_1/bin/gcc-9 -std=c++17 -c test.cc 
In file included from /usr/local/Cellar/gcc/9.3.0_1/include/c++/9.3.0/stdlib.h:36,
                 from test.cc:6:
/usr/local/Cellar/gcc/9.3.0_1/include/c++/9.3.0/cstdlib:132:11: error: '::aligned_alloc' has not been declared
  132 |   using ::aligned_alloc;
      |           ^~~~~~~~~~~~~
```

ax3l commented:
@alalazo Assuming you re-build your Spack-provided GCC, taking this diff [as in homebrew](https://github.com/Homebrew/homebrew-core/pull/45585) ([more background](https://gist.github.com/McMartin/6a4d07baa8e7fd45b5c864cb67c12a9c)) into account:
```diff
diff --git a/lib/spack/spack/operating_systems/mac_os.py b/lib/spack/spack/operating_systems/mac_os.py
index 519820f93..38df0aa05 100644
--- a/lib/spack/spack/operating_systems/mac_os.py
+++ b/lib/spack/spack/operating_systems/mac_os.py
@@ -21,7 +21,7 @@ def macos_sdk_path():
     """Return SDK path
     """
     xcrun = Executable('xcrun')
-    return xcrun('--show-sdk-path', output=str, error=str).rstrip()
+    return xcrun('--sdk macosx --show-sdk-path', output=str, error=str).rstrip()
 
 
 class MacOs(OperatingSystem):
```
Does this fix your Spack-GCC powered CMake build or [your reproducer](https://gitlab.kitware.com/cmake/cmake/-/issues/21135#note_820864)?
```c++
#if !defined(_WIN32) && !defined(__sun)
/* POSIX APIs are needed */
#  define _POSIX_C_SOURCE 200809L
#endif

#include <stdlib.h>

int main(){}
```

Looks like this landed already in GCC 9.2 rev2 in Homebrew, so I don't get why @gartung get's this with his homebrew GCC 9.3.0 as well. On the other hand, things changed again in homebrew (back to the old broken sdk hint?):
https://github.com/Homebrew/homebrew-core/blob/1b68f13d7d1efd3725280e0c30a8948d75141e75/Formula/gcc.rb#L83
